### DISCUSSION TITLE
Plans for caching?

### DISCUSSION THREAD

**Discussion by JoolsMcFly** [VincentLanglet/Twig-CS-Fixer#60] [Ideas]
Hi,

thanks for this cool fixer. 
It currently takes 13s to fix our templates folder and I was wondering if there was any plans to add a caching layer like PHP-CS-Fixer has?

Thanks!

**Comment by VincentLanglet** [comment]
Hi. Caching would be a great feature indeed.
I didn't look at how it was made in PHP-CS-Fixer, but we could try to do similar things.

Do you want to try ?

**Reply by JoolsMcFly** [reply]
I'd like to! Problem is finding time.
I'll have a closer look at PHP-CS-FIXER implementation and see if I can port it here.

**Reply by JoolsMcFly** [reply]
Hey @VincentLanglet 

I took some time to work on this today and I have a working POC even. I still need to fix tests in `TwigCsFixerCommandTest.php`. We might not need ToolInfo if we were to move code into Config for instance? Any ideas welcome.

Not sure what to do with `@internal` in `/Cache` folder. Leave it?

Also, since it's heavily based on PHP-CS-Fixer, should we add a readme in the `Cache` folder or directly in this project's readme?

Anyway, feel free to review my changes here: https://github.com/VincentLanglet/Twig-CS-Fixer/pull/66 (it's still a draft)

Thanks.

**Reply by VincentLanglet** [reply]
Nice start. I've made some reviews

For ToolInfo, I think Composer/InstalledVersions class should already give things you need.
For `@internal` it's not needed, you can remove them.

**Reply by JoolsMcFly** [reply]
Sounds good. I’ll make suggested changes tomorrow.

**Reply by JoolsMcFly** [reply]
Next step expected from me: add some tests I suppose?

**Reply by VincentLanglet** [reply]
I made some minor reviews but yes we need some tests.
I assume you already tried this branch on your project and it works ?

I'll be able to help writing tests if you need to.

**Reply by JoolsMcFly** [reply]
I'll be AFK until Monday so I'll run more tests manually then. When done I'll get back to you and see what we can do re tests. Sounds good?

**Reply by JoolsMcFly** [reply]
Hey @VincentLanglet.

I pushed some changes and run tests locally: with/without user defined cache, files that need or need not fixing etc.
I can go ahead and see about tests if it's all good on your end.

**Reply by VincentLanglet** [reply]
If it works for you and tests are working, I think we can merge this like this.
I'll be able to write tests on a separate PR and it will be easier to try the `dev-main` branch on some personal projects.

**Reply by JoolsMcFly** [reply]
OK, I added some basic tests for reading/writing cache files and refactored according to your latest suggestions.

If I add a call to `setCacheFile` in `.twig-cs-fixer.php` then custom cache file is used. Otherwise it defaults to `.twig-cs-fixer.cache`.

I tested adding / removing sniffs from `.twig-cs-fixer.php` and cache was invalidated accordingly.
If I don't make changes to a file after it was fixed then it's skipped. If content is different then it is parsed.

If you feel like testing a bit on your end then feel free! In the meantime I'll remove the draft status from my PR.

**Reply by VincentLanglet** [reply]
@JoolsMcFly I have a last issue I'd like to solve before the release.

I try the cache feature with
```
$config = new Config();
$config->setCacheFile('.build/twig-cs-fixer/.twig-cs-fixer.cache');
```
when the twig-cs-fixer directory does not exist.

I got the error `Directory of cache file "%s" does not exists.`.
I'd like the directory to be created instead, in the same time than the file.

What I don't understand is the fact this check exists in PHP-CS-Fixer, and I don't have the error
https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/src/Cache/FileHandler.php#L80

Any idea ?

**Reply by VincentLanglet** [reply]
My bad, it seems like I had
```
if (!file_exists($dirCache)) {
    mkdir($dirCache, 0777, true);
}
```
in my php-cs-fixer config, so I have the same behavior.

So it seems ready to be released.

**Reply by JoolsMcFly** [reply]
Phew. ;)
All good then!
